src/kv/untyped_map.h

```diff
@@ -0,0 +1,734 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ds/dl_list.h"
+#include "ds/logger.h"
+#include "ds/spin_lock.h"
+#include "kv/kv_serialiser.h"
+#include "kv/kv_types.h"
+#include "kv/untyped_tx_view.h"
+
+#include <functional>
+#include <optional>
+#include <unordered_set>
+
+namespace kv::untyped
+{
+  namespace Check
+  {
+    struct No
+    {};
+
+    template <typename T, typename Arg>
+    No operator!=(const T&, const Arg&)
+    {
+      return No();
+    }
+
+    template <typename T, typename Arg = T>
+    struct Ne
+    {
+      enum
+      {
+        value = !std::is_same<decltype(*(T*)(0) != *(Arg*)(0)), No>::value
+      };
+    };
+
+    template <class T>
+    bool ne(std::enable_if_t<Ne<T>::value, const T&> a, const T& b)
+    {
+      return a != b;
+    }
+
+    template <class T>
+    bool ne(std::enable_if_t<!Ne<T>::value, const T&> a, const T& b)
+    {
+      return false;
+    }
+  }
+
+  struct LocalCommit
+  {
+    LocalCommit() = default;
+    LocalCommit(Version v, State&& s, Write&& w) :
+      version(v),
+      state(std::move(s)),
+      writes(std::move(w))
+    {}
+    LocalCommit(Version v, const State& s, const Write& w) :
+      version(v),
+      state(s),
+      writes(w)
+    {}
+
+    Version version;
+    State state;
+    Write writes;
+    LocalCommit* next = nullptr;
+    LocalCommit* prev = nullptr;
+  };
+  using LocalCommits = snmalloc::DLList<LocalCommit, std::nullptr_t, true>;
+
+  struct Roll
+  {
+    std::unique_ptr<LocalCommits> commits;
+    size_t rollback_counter;
+  };
+
+  struct RollContainer
```

Comment:
Template dependent lookup let us get away with some circular dependencies between `kv::Map<K, V>` and `kv::TxViewCommitter<K, V>`. This doesn't work now they're non-templated, `kv::Map` and `kv::untyped::TxView>`. To break the loop we create this little interface, which is how a `TxViewCommitter` sees its parent `Map`. No more friendship, just a vague `RollContainer`.